sdks/java/harness/src/main/java/org/apache/beam/fn/harness/logging/BeamFnLoggingClient.java

```diff
@@ -0,0 +1,322 @@
+    private final BlockingDeque<BeamFnApi.LogEntry> bufferedLogEntries =
+        new LinkedBlockingDeque<>(MAX_BUFFER_SIZE);
+    private final Future<?> bufferedLogWriter;
+    private final ThreadLocal<Consumer<BeamFnApi.LogEntry>> logEntryHandler;
+
+    private LogRecordHandler(ExecutorService executorService) {
+      bufferedLogWriter = executorService.submit(this);
+      logEntryHandler = new ThreadLocal<>();
+    }
+
+    @Override
+    public void publish(LogRecord record) {
+      BeamFnApi.LogEntry.Severity severity = LOG_LEVEL_MAP.get(record.getLevel());
+      if (severity == null) {
+        return;
+      }
+      BeamFnApi.LogEntry.Builder builder = BeamFnApi.LogEntry.newBuilder()
+          .setSeverity(severity)
+          .setLogLocation(record.getLoggerName())
+          .setMessage(FORMATTER.formatMessage(record))
+          .setThread(Integer.toString(record.getThreadID()))
+          .setTimestamp(Timestamp.newBuilder()
+              .setSeconds(record.getMillis() / 1000)
+              .setNanos((int) (record.getMillis() % 1000) * 1_000_000));
+      if (record.getThrown() != null) {
+        builder.setTrace(formatException(record.getThrown()));
+      }
+      // The thread that sends log records should never perform a blocking publish and
+      // only insert log records best effort. We detect which thread is logging
+      // by using the thread local, defaulting to the blocking publish.
+      MoreObjects.firstNonNull(
+          logEntryHandler.get(), this::blockingPublish).accept(builder.build());
+    }
+
+    /** Blocks caller till enough space exists to publish this log entry. */
+    private void blockingPublish(BeamFnApi.LogEntry logEntry) {
+      try {
+        bufferedLogEntries.put(logEntry);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public void run() {
+      // Logging which occurs in this thread will attempt to publish log entries into the
+      // above handler which should never block if the queue is full otherwise
+      // this thread will get stuck.
+      logEntryHandler.set(bufferedLogEntries::offer);
+      List<BeamFnApi.LogEntry> additionalLogEntries = new ArrayList<>(MAX_BUFFER_SIZE);
+      try {
+        BeamFnApi.LogEntry logEntry;
+        while ((logEntry = bufferedLogEntries.take()) != POISON_PILL) {
+          BeamFnApi.LogEntry.List.Builder builder =
+              BeamFnApi.LogEntry.List.newBuilder().addLogEntries(logEntry);
+          bufferedLogEntries.drainTo(additionalLogEntries);
+          for (int i = 0; i < additionalLogEntries.size(); ++i) {
+            if (additionalLogEntries.get(i) == POISON_PILL) {
+              additionalLogEntries = additionalLogEntries.subList(0, i);
+              break;
+            }
+          }
+          builder.addAllLogEntries(additionalLogEntries);
```

Comment:
The complexity is that I don't want to enter the lock of the queue more than once which is why I add them all to this other list. I could add them one by one but I think the bulk add will be faster since it will be looping inside the proto code. Probably doesn't really matter.